### ISSUE TITLE
Validates uniqueness on a belongs_to association doesn't use foreign_key value.

### ISSUE THREAD

**Issue by joshuapinter** [rails/rails#49293]
### Steps to reproduce

When you have a `belongs_to` association that uses a custom `foreign_key`, using `validates` with `uniqueness` uses the `id` column value instead of the `foreign_key` column value. To explain, here is a simplified setup:

```ruby
class StudentCourse
  belongs_to :student, primary_key: :uid, foreign_key: :student_uid
  belongs_to :course,  primary_key: :uid, foreign_key: :course_uid

  validates :student, presence: true, uniqueness: { scope: :course, message: "is already in the Course" }
end
```

And assume there is one `Student` record and one `Course` record each with an `id` of 1 and a `uid` of "A".

And then look at the logs and trigger the validation query with:

```ruby
student_course.valid?
```

### Expected behavior
```sql
StudentCourse Exists? (0.7ms)  SELECT 1 AS one FROM `course_students` WHERE `course_students`.`student_uid` = 'A' AND `course_students`.`course_uid` = 'A' LIMIT 1
```

### Actual behavior
```sql
StudentCourse Exists? (0.7ms)  SELECT 1 AS one FROM `course_students` WHERE `course_students`.`student_uid` = '1' AND `course_students`.`course_uid` = 'A' LIMIT 1
```

Notice that it is using the value of 1 for `student_uid` instead of `"A"`.

### Additional Investigation
NOTE: If I instead use the `foreign_key` column (`student_uid`) directly in the validation instead of the association, like: 
```ruby
validates :student_uid, presence: true, uniqueness: { scope: :course, message: "is already in the Course" }
```

Then you get the expected behaviour of:

```sql
StudentCourse Exists? (0.7ms)  SELECT 1 AS one FROM `course_students` WHERE `course_students`.`student_uid` = 'A' AND `course_students`.`course_uid` = 'A' LIMIT 1
```

So the issue appears to be with the association and how Rails is ignoring the `foreign_key` in the association definition.

### System configuration
**Rails version**: 7.0.7

**Ruby version**: ruby 3.2.2 (2023-03-30 revision e51014f9c0) [arm64-darwin22]

**Comment by fatkodima** [comment]
Please create a reproduction script using https://github.com/rails/rails/blob/main/guides/bug_report_templates/active_record_gem.rb template.

**Comment by joshuapinter** [comment]
@fatkodima Cool, is this just run in the console? Or how do I got about testing the template before posting here?

**Comment by fatkodima** [comment]
Store it locally and run as a ruby file. This can help - https://guides.rubyonrails.org/contributing_to_ruby_on_rails.html#creating-a-bug-report.

**Comment by joshuapinter** [comment]
@fatkodima Here is the test script: 

```ruby
# frozen_string_literal: true

require 'bundler/inline'

gemfile(true) do
  source 'https://rubygems.org'

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  # Activate the gem you are reporting the issue against.
  gem 'activerecord', '~> 7.0.7'
  gem 'sqlite3'
end

require 'active_record'
require 'minitest/autorun'
require 'logger'

# This connection will do for database-independent bug reports.
ActiveRecord::Base.establish_connection(
  adapter: 'sqlite3',
  database: ':memory:'
)
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :courses, force: true do |t|
    t.string :uid
  end

  create_table :students, force: true do |t|
    t.string :uid
  end

  create_table :student_courses, force: true do |t|
    t.string :course_uid
    t.string :student_uid
  end
end

class Course < ActiveRecord::Base
  has_many :students, through: :student_courses
end

class Student < ActiveRecord::Base
  has_many :courses, through: :student_courses
end

class StudentCourse < ActiveRecord::Base
  belongs_to :course,  primary_key: :uid, foreign_key: :course_uid
  belongs_to :student, primary_key: :uid, foreign_key: :student_uid

  validates :student, presence: true, uniqueness: { scope: :course, message: 'is already in the Course' }
end

class BugTest < Minitest::Test
  def test_student_unique_for_course
    Course.create!(uid: 'A' )
    Student.create!(uid: 'A' )
    StudentCourse.create!( course_uid: 'A', student_uid: 'A' )
    duplicate_student_course = StudentCourse.new( course_uid: 'A', student_uid: 'A' )
    refute duplicate_student_course.valid?
  end
end
```

And when running it the log file you're looking for which confirms the issue is the following: 

```sql
StudentCourse Exists? (0.0ms)  SELECT 1 AS one FROM "student_courses" WHERE "student_courses"."student_uid" = ? AND "student_courses"."course_uid" = ? LIMIT ?  [["student_uid", "1"], ["course_uid", "A"], ["LIMIT", 1]]
```

Again, you can see that it's using the Student's `id` value instead of the `student_uid` value.

Let me know if you need anything else. 

Thanks!

**Comment by ghiculescu** [comment]
Thanks for the issue. Would you like to propose a PR to fix it?

I'm guessing the fix will be in https://github.com/rails/rails/blob/d21d811ffece4d3959bcd37e58fec77590ff6f93/activerecord/lib/active_record/validations/uniqueness.rb#L111, and you could add a test to https://github.com/rails/rails/blob/main/activerecord/test/cases/validations/uniqueness_validation_test.rb

**Comment by maniSHarma7575** [comment]
@ghiculescu @joshuapinter

This issue has already been resolved in Rails 7.1.0.beta1 thanks to PR #45790.

Here is the test script:

```
# frozen_string_literal: true

require 'bundler/inline'

gemfile(true) do
  source 'https://rubygems.org'

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  # Activate the gem you are reporting the issue against.
  gem 'activerecord', '~> 7.1.0.beta1'
  gem 'sqlite3'
end

require 'active_record'
require 'minitest/autorun'
require 'logger'

# This connection will do for database-independent bug reports.
ActiveRecord::Base.establish_connection(
  adapter: 'sqlite3',
  database: ':memory:'
)
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :courses, force: true do |t|
    t.string :uid
  end

  create_table :students, force: true do |t|
    t.string :uid
  end

  create_table :student_courses, force: true do |t|
    t.string :course_uid
    t.string :student_uid
  end
end

class Course < ActiveRecord::Base
  has_many :students, through: :student_courses
end

class Student < ActiveRecord::Base
  has_many :courses, through: :student_courses
end

class StudentCourse < ActiveRecord::Base
  belongs_to :course,  primary_key: :uid, foreign_key: :course_uid
  belongs_to :student, primary_key: :uid, foreign_key: :student_uid

  validates :student, presence: true, uniqueness: { scope: :course, message: 'is already in the Course' }
end

class BugTest < Minitest::Test
  def test_student_unique_for_course
    Course.create!(uid: 'A' )
    Student.create!(uid: 'A' )
    StudentCourse.create!( course_uid: 'A', student_uid: 'A' )
    duplicate_student_course = StudentCourse.new( course_uid: 'A', student_uid: 'A' )
    refute duplicate_student_course.valid?
  end
end
```

Test script output:

```
Fetching gem metadata from https://rubygems.org/........
Resolving dependencies...
-- create_table(:courses, {:force=>true})
D, [2023-09-16T14:29:35.386697 #99083] DEBUG -- :    (0.0ms)  DROP TABLE IF EXISTS "courses"
D, [2023-09-16T14:29:35.386910 #99083] DEBUG -- :    (0.1ms)  CREATE TABLE "courses" ("id" integer PRIMARY KEY AUTOINCREMENT NOT NULL, "uid" varchar)
   -> 0.0056s
-- create_table(:students, {:force=>true})
D, [2023-09-16T14:29:35.387021 #99083] DEBUG -- :    (0.0ms)  DROP TABLE IF EXISTS "students"
D, [2023-09-16T14:29:35.387090 #99083] DEBUG -- :    (0.0ms)  CREATE TABLE "students" ("id" integer PRIMARY KEY AUTOINCREMENT NOT NULL, "uid" varchar)
   -> 0.0002s
-- create_table(:student_courses, {:force=>true})
D, [2023-09-16T14:29:35.387178 #99083] DEBUG -- :    (0.0ms)  DROP TABLE IF EXISTS "student_courses"
D, [2023-09-16T14:29:35.387246 #99083] DEBUG -- :    (0.0ms)  CREATE TABLE "student_courses" ("id" integer PRIMARY KEY AUTOINCREMENT NOT NULL, "course_uid" varchar, "student_uid" varchar)
   -> 0.0001s
D, [2023-09-16T14:29:35.387676 #99083] DEBUG -- :    (0.0ms)  CREATE TABLE "schema_migrations" ("version" varchar NOT NULL PRIMARY KEY)
D, [2023-09-16T14:29:35.388537 #99083] DEBUG -- :    (0.0ms)  CREATE TABLE "ar_internal_metadata" ("key" varchar NOT NULL PRIMARY KEY, "value" varchar, "created_at" datetime(6) NOT NULL, "updated_at" datetime(6) NOT NULL)
D, [2023-09-16T14:29:35.396454 #99083] DEBUG -- :   ActiveRecord::InternalMetadata Load (0.0ms)  SELECT * FROM "ar_internal_metadata" WHERE "ar_internal_metadata"."key" = ? ORDER BY "ar_internal_metadata"."key" ASC LIMIT 1  [[nil, "environment"]]
D, [2023-09-16T14:29:35.396581 #99083] DEBUG -- :   ActiveRecord::InternalMetadata Create (0.0ms)  INSERT INTO "ar_internal_metadata" ("key", "value", "created_at", "updated_at") VALUES ('environment', 'default_env', '2023-09-16 08:59:35.396483', '2023-09-16 08:59:35.396485')
Run options: --seed 53707

# Running:

D, [2023-09-16T14:29:35.409246 #99083] DEBUG -- :   TRANSACTION (0.0ms)  begin transaction
D, [2023-09-16T14:29:35.409367 #99083] DEBUG -- :   Course Create (0.2ms)  INSERT INTO "courses" ("uid") VALUES (?)  [["uid", "A"]]
D, [2023-09-16T14:29:35.409465 #99083] DEBUG -- :   TRANSACTION (0.0ms)  commit transaction
D, [2023-09-16T14:29:35.410465 #99083] DEBUG -- :   TRANSACTION (0.0ms)  begin transaction
D, [2023-09-16T14:29:35.410501 #99083] DEBUG -- :   Student Create (0.0ms)  INSERT INTO "students" ("uid") VALUES (?)  [["uid", "A"]]
D, [2023-09-16T14:29:35.410546 #99083] DEBUG -- :   TRANSACTION (0.0ms)  commit transaction
D, [2023-09-16T14:29:35.416796 #99083] DEBUG -- :   TRANSACTION (0.0ms)  begin transaction
D, [2023-09-16T14:29:35.416852 #99083] DEBUG -- :   Student Load (0.1ms)  SELECT "students".* FROM "students" WHERE "students"."uid" = ? LIMIT ?  [["uid", "A"], ["LIMIT", 1]]
D, [2023-09-16T14:29:35.417237 #99083] DEBUG -- :   Course Load (0.0ms)  SELECT "courses".* FROM "courses" WHERE "courses"."uid" = ? LIMIT ?  [["uid", "A"], ["LIMIT", 1]]
D, [2023-09-16T14:29:35.417692 #99083] DEBUG -- :   StudentCourse Exists? (0.0ms)  SELECT 1 AS one FROM "student_courses" WHERE "student_courses"."student_uid" = ? AND "student_courses"."course_uid" = ? LIMIT ?  [["student_uid", "A"], ["course_uid", "A"], ["LIMIT", 1]]
D, [2023-09-16T14:29:35.417839 #99083] DEBUG -- :   StudentCourse Create (0.0ms)  INSERT INTO "student_courses" ("course_uid", "student_uid") VALUES (?, ?)  [["course_uid", "A"], ["student_uid", "A"]]
D, [2023-09-16T14:29:35.417903 #99083] DEBUG -- :   TRANSACTION (0.0ms)  commit transaction
D, [2023-09-16T14:29:35.418197 #99083] DEBUG -- :   Student Load (0.0ms)  SELECT "students".* FROM "students" WHERE "students"."uid" = ? LIMIT ?  [["uid", "A"], ["LIMIT", 1]]
D, [2023-09-16T14:29:35.418346 #99083] DEBUG -- :   Course Load (0.0ms)  SELECT "courses".* FROM "courses" WHERE "courses"."uid" = ? LIMIT ?  [["uid", "A"], ["LIMIT", 1]]
D, [2023-09-16T14:29:35.418491 #99083] DEBUG -- :   StudentCourse Exists? (0.0ms)  SELECT 1 AS one FROM "student_courses" WHERE "student_courses"."student_uid" = ? AND "student_courses"."course_uid" = ? LIMIT ?  [["student_uid", "A"], ["course_uid", "A"], ["LIMIT", 1]]
.

Finished in 0.013867s, 72.1137 runs/s, 72.1137 assertions/s.

1 runs, 1 assertions, 0 failures, 0 errors, 0 skips
```
